### PR TITLE
refactor(lsp): simplify client tracking

### PR DESCRIPTION
**Pull request by lewis6991**
- Remove:
    - uninitialized_clients
    - active_clients
    - all_buffer_active_clients
- Add:
    - all_clients

- Use `lsp.get_clients()` to get buffer clients.

### CODE REVIEW

**Review by mfussenegger — APPROVED** [review]
Very nice :+1:

**Comment by lewis6991** [comment]
Turns out the client tracking is very delicate and some of the tests aren't quite right. This take some more time to straighten things out.

File: test/functional/plugin/lsp_spec.lua [hunk 1]
```diff
@@ -806,14 +808,12 @@ describe('LSP', function()
...
               vim.lsp.handlers['textDocument/typeDefinition'] = function() end
+              vim.cmd(BUFFER.."bwipeout")
             ]=])
         end,
         on_init = function(client)
           client.stop()
           exec_lua('vim.lsp.buf.type_definition()')
-          exec_lua [[
-            vim.api.nvim_command(BUFFER.."bwipeout")
-          ]]
```
Comment (lewis6991) [inline test/functional/plugin/lsp_spec.lua:816]: Deleting a buffer after the client is initialized should cause the client to call `didOpen` and `didClose` which didn't happen before. Moving the buffer deletion earlier prevents this.

File: test/functional/plugin/lsp_spec.lua [hunk 2]
```diff
@@ -1058,8 +1058,7 @@ describe('LSP', function()
       local client
       test_rpc_server {
         test_name = 'basic_finish',
-        on_init = function(_client)
-          client = _client
```
Comment (lewis6991) [inline test/functional/plugin/lsp_spec.lua:1062]: Again attaching a client to a buffer in `on_init` (which happens after it has initialized) should cause it to call `didOpen`.

**Pull request merged**